README.rst
|Release| |PyPI| |MIT licensed| |ASCL| |Travis Build| |Issues| |RTD| |Landscape| 

Description
===========

This repo contains a set of codes to measure the following OpenMP
parallelized clustering measures in a cosmological box (co-moving XYZ)
or on a mock (RA, DEC, CZ). Also, contains the associated paper to be
published in Astronomy & Computing Journal (at some point). Read the
documentation on `corrfunc.rtfd.io <http://corrfunc.rtfd.io/>`_. 

Why Should You Use it
======================

1. **Fast** Theory pair-counting is **7x** faster than ``SciPy cKDTree``, and at least **2x** faster than all existing public codes.
2. **OpenMP Parallel** All pair-counting codes can be done in parallel (with strong scaling efficiency >~ 95% up to 10 cores)
3. **Python Extensions** Python extensions allow you to do the compute-heavy bits using C while retaining all of the user-friendliness of python. 
4. **Weights** All correlation functions now support weights for individual points (in ``master`` branch, upcoming in `v2.0.0 <https://github.com/manodeep/corrfunc/releases/tag/2.0.0>`_)
5. **Modular** The code is written in a modular fashion and is easily extensible to compute arbitrary clustering statistics. 
6. **Future-proof** As I get access to newer instruction-sets, the codes will get updated to use the latest and greatest CPU features. 

*If you use the codes for your analysis, please star this repo -- that helps us keep track of the number of users.*

Benchmark against Existing Codes
================================

Please see this
`gist <https://gist.github.com/manodeep/cffd9a5d77510e43ccf0>`__ for
some benchmarks with current codes. If you have a pair-counter that you would like to compare, please add in a corresponding function and update the timings. 

Installation
============

Pre-requisites
--------------

1. ``make >= 3.80``
2. OpenMP capable compiler like ``icc``, ``gcc>=4.6`` or ``clang >= 3.7``. If
   not available, please disable ``USE_OMP`` option option in
   ``theory.options`` and ``mocks.options``. You might need to ask your
   sys-admin for system-wide installs of the compiler; if you prefer to
   install your own then ``conda install gcc`` (MAC/linux) or
   ``(sudo) port install gcc5`` (on MAC) should work. 
3. ``gsl``. Use either
   ``conda install -c https://conda.anaconda.org/asmeurer gsl``
   (MAC/linux) or ``(sudo) port install gsl`` (MAC) to install ``gsl``
   if necessary.
4. ``python >= 2.6`` or ``python>=3.4`` for compiling the C extensions.
5. ``numpy>=1.7`` for compiling the C extensions.

Preferred Install Method
-------------------------

::

    $ git clone https://github.com/manodeep/corrfunc/
    $ make 
    $ make install
    $ python setup.py install (--user)
    $ make tests 

Assuming you have ``gcc`` in your ``PATH``, ``make`` and
``make install`` should compile and install the C libraries + python
extensions within the source directory. If you would like to install the
python C extensions in your environment, then
``python setup.py install (--user)`` should be sufficient. If you are primarily
interested in the ``python`` interface, you can condense all of the steps
by using ``python setup.py install CC=yourcompiler (--user)`` after ``git clone``.

Compilation Notes
------------------

- If python and/or numpy are not available, then the C extensions will not be compiled.

- Default compiler on MAC is set to ``clang``, if you want to specify a different compiler, you will have to call ``make CC=yourcompiler``


Alternate Install Method
-------------------------

The python package is directly installable via ``pip install corrfunc``. However, in that case you will lose the ability to recompile the code according to your needs. Installing via `pip` is **not** recommended, please open an install issue on this repo first; doing so helps improve the code-base and saves future users from running into similar install issues. 

Installation notes
------------------

If compilation went smoothly, please run ``make tests`` to ensure the
code is working correctly. Depending on the hardware and compilation
options, the tests might take more than a few minutes. *Note that the
tests are exhaustive and not traditional unit tests*.

While I have tried to ensure that the package compiles and runs out of
the box, cross-platform compatibility turns out to be incredibly hard.
If you run into any issues during compilation and you have all of the
pre-requisites, please see the `FAQ <FAQ>`__ or `email
the corrfunc mailing list <mailto:corrfunc@googlegroups.com>`__. Also, feel free to create a new issue
with the ``Installation`` label.

Clustering Measures on a Cosmological box
-----------------------------------------

All codes that work on cosmological boxes with co-moving positions are
located in the ``theory`` directory. The various clustering measures
are:

1. ``DD`` -- Measures auto/cross-correlations between two boxes.
   The boxes do not need to be cubes.

2. ``xi`` -- Measures 3-d auto-correlation in a cubic cosmological box.
   Assumes PERIODIC boundary conditions.

3. ``wp`` -- Measures auto 2-d point projected correlation function in a
   cubic cosmological box. Assumes PERIODIC boundary conditions.

4. ``DDrppi`` -- Measures the auto/cross correlation function between
   two boxes. The boxes do not need to be cubes.

5. ``vpf`` -- Measures the void probability function + counts-in-cells.

Clustering measures on a Mock
-----------------------------

All codes that work on mock catalogs (RA, DEC, CZ) are located in the
``mocks`` directory. The various clustering measures are:

1. ``DDrppi_mocks`` -- The standard auto/cross correlation between two data
   sets. The outputs, DD, DR and RR can be combined using ``wprp`` to
   produce the Landy-Szalay estimator for `wp(rp)`.

2. ``DDtheta_mocks`` -- Computes angular correlation function between two data
   sets. The outputs from ``DDtheta_mocks`` need to be combined with
   ``wtheta`` to get the full `\omega(\theta)`

3. ``vpf_mocks`` -- Computes the void probability function on mocks.

Science options
===============

If you plan to use the command-line, then you will have to specify the
code runtime options at compile-time. For theory routines, these options
are in the file ``theory.options`` while for the mocks, these options are
in file ``mocks.options``. 

**Note** All options can be specified at 
runtime if you use the python interface or the static libraries. Each one of
the following ``Makefile`` option has a corresponding entry for the runtime
libraries. 

Theory (in ``theory.options``)
-------------------------------

1. ``PERIODIC`` (ignored in case of wp/xi) -- switches periodic boundary
   conditions on/off. Enabled by default.

2. ``OUTPUT_RPAVG`` -- switches on output of ``<rp>`` in each ``rp``
   bin. Can be a massive performance hit (~ 2.2x in case of wp).
   Disabled by default. 

3. ``DOUBLE_PREC`` -- switches on calculations in double precision. Disabled
   by default (i.e., calculations are performed in single precision by default).
   
Mocks (in ``mocks.options``)
----------------------------

1. ``OUTPUT_RPAVG`` -- switches on output of ``<rp>`` in each ``rp``
   bin for ``DDrppi_mocks``. Enabled by default.

2. ``OUTPUT_THETAAVG`` -- switches on output of in each theta bin. Can
   be extremely slow (~5x) depending on compiler, and CPU capabilities.
   Disabled by default.

3. ``DOUBLE_PREC`` -- switches on calculations in double precision. Disabled
   by default (i.e., calculations are performed in single precision by default).
   
4. ``LINK_IN_DEC`` -- creates binning in declination for ``DDtheta``. Please
   check that for your desired limits ``\theta``, this binning does not 
   produce incorrect results (due to numerical precision). Generally speaking,
   if your ``\thetamax`` (the max. ``\theta`` to consider pairs within) is too
   small (probaly less than 1 degree), then you should check with and without
   this option. Errors are typically sub-percent level. 

5. ``LINK_IN_RA`` -- creates binning in RA once binning in DEC has been
   enabled. Same numerical issues as ``LINK_IN_DEC``

6. ``FAST_DIVIDE`` -- Disabled by default. Divisions are slow but required
   ``DD(r_p,\pi)``. Enabling this option, replaces
   the divisions with a reciprocal followed by a Newton-Raphson. The code
   will run ~20% faster at the expense of some numerical precision.
   Please check that the loss of precision is not important for your
   use-case. 

7. ``FAST_ACOS`` -- Relevant only when ``OUTPUT_THETAAVG`` is enabled. Disabled 
   by default. An ``arccos`` is required to calculate ``<\theta>``. In absence of vectorized
   ``arccos`` (intel compiler, ``icc`` provides one via intel Short Vector Math 
   Library), this calculation is extremely slow. However, we can approximate
   ``arccos`` using polynomials (with `Remez Algorithm <https://en.wikipedia.org/wiki/Remez_algorithm>`_).
   The approximations are taken from implementations released by `Geometric Tools <http://geometrictools.com/>`_.
   Depending on the level of accuracy desired, this implementation of ``fast acos`` 
   can be tweaked in the file `utils/fast_acos.h <utils/fast_acos.h>`__. An alternate, less
   accurate implementation is already present in that file. Please check that the loss of 
   precision is not important for your use-case. 

8. ``COMOVING_DIST`` -- Currently there is no support in ``corrfunc`` for different cosmologies. However, for the
   mocks routines like, ``DDrppi_mocks`` and ``vpf_mocks``, cosmology parameters are required to convert between
   redshift and co-moving distance. Both ``DDrppi_mocks`` and ``vpf_mocks`` expects to receive a ``redshift`` array 
   as input; however, with this option enabled, the ``redshift`` array will be assumed to contain already converted
   co-moving distances. So, if you have redshifts and want to use an arbitrary cosmology, then convert the redshifts
   into co-moving distances, enable this option, and pass the co-moving distance array into the routines. 

Running the codes
=================

Read the documentation on `corrfunc.rtfd.io <http://corrfunc.rtfd.io/>`_.


Using the command-line interface
--------------------------------

Navigate to the correct directory. Make sure that the options, set in
either ``theory.options`` or ``mocks.options`` in the root directory are
what you want. If not, edit those two files (and possibly
``common.mk``), and recompile. Then, you can use the command-line
executables in each individual subdirectory corresponding to the
clustering measure you are interested in. For example, if you want to
compute the full 3-D correlation function, ``\xi(r)``, then navigate to
``theory/xi`` and run the executable ``xi``. If you run executables
without any arguments, the message will you tell you all the required
arguments.

Calling from C
--------------

Look under the ``theory/examples/run_correlations.c`` and
``mocks/examples/run_correlations_mocks.c`` to see examples of
calling the C API directly. If you run the executables,
``run_correlations`` and ``run_correlations_mocks``, the output will
also show how to call the command-line interface for the various
clustering measures.

Calling from Python
-------------------

If all went well, the codes can be directly called from ``python``.
Please see ``corrfunc/call_correlation_functions.py`` and
``corrfunc/call_correlation_functions_mocks.py`` for examples on how to
use the C extensions directly. Here are a few examples:

.. code:: python

    from __future__ import print_function
    import os.path as path
    import numpy as np
    import corrfunc
    from corrfunc.theory import wp

    # Setup the problem for wp
    boxsize = 500.0
    pimax = 40.0
    nthreads = 4

    # Create a fake data-set.
    Npts = 100000
    x = np.float32(np.random.random(Npts))
    y = np.float32(np.random.random(Npts))
    z = np.float32(np.random.random(Npts))
    x *= boxsize
    y *= boxsize
    z *= boxsize

    # Setup the bins
    rmin = 0.1
    rmax = 20.0
    nbins = 20
    
    # Create the bins
    rbins = np.logspace(np.log10(0.1), np.log10(rmax), nbins + 1)

    # Call wp
    wp_results = wp(boxsize, pimax, nthreads, rbins, x, y, z, verbose=True, output_rpavg=True)

    # Print the results
    print("#############################################################################")
    print("##       rmin           rmax            rpavg             wp            npairs")
    print("#############################################################################")
    print(wp_results)
                                                        

Common Code options for both Mocks and Cosmological Boxes
=========================================================

1. ``USE_OMP`` -- uses OpenMP parallelization. Scaling is great for DD
   (perfect scaling up to 12 threads in my tests) and okay (runtime
   becomes constant ~6-8 threads in my tests) for ``DDrppi`` and ``wp``.
   Enabled by default. The ``Makefile`` will compare the `CC` variable with
   known OpenMP enabled compilers and set compile options accordingly. 
   Set in ``common.mk`` by default. 

*Optimization for your architecture*

1. The values of ``bin_refine_factor`` and/or ``zbin_refine_factor`` in
   the ``countpairs\_\*.c`` files control the cache-misses, and
   consequently, the runtime. In my trial-and-error methods, I have seen
   any values larger than 3 are always slower. But some different
   combination of 1/2 for ``(z)bin_refine_factor`` might be faster on
   your platform.

2. If you have AVX2/AVX-512/KNC, you will need to add a new kernel within
   the ``*_kernels.c`` and edit the runtime dispatch code to call this new
   kernel. 

Author & Maintainers 
=====================

corrfunc was designed by Manodeep Sinha and is currently maintained by
`Lehman Garrison <https://github.com/lgarrison>`_ and `Manodeep Sinha <https://github.com/manodeep>`_

Citing
======

If you use the code, please cite using the `ascl entry <http://ascl.net/1703.003>`_ as indexed by `ADS <http://adsabs.harvard.edu/abs/2017ascl.soft03003S>`_. The BibTex entry for the code is

::

         @misc{2017ascl.soft03003S,
             author = {{Sinha}, M. and {Garrison}, L.},
             title = "{Corrfunc: Blazing fast correlation functions on the CPU}",
             keywords = {Software},
             howpublished = {Astrophysics Source Code Library},
             year = 2017,
             month = mar,
             archivePrefix = "ascl",
             eprint = {1703.003},
             adsurl = {http://adsabs.harvard.edu/abs/2017ascl.soft03003S},
             adsnote = {Provided by the SAO/NASA Astrophysics Data System}
         }

         
Mailing list
============

If you have questions or comments about the package, please do so on the
mailing list: https://groups.google.com/forum/#!forum/corrfunc

LICENSE
=======

corrfunc is released under the MIT license. Basically, do what you want
with the code including using it in commercial application.

Project URL
===========

-  website (https://manodeep.github.io/Corrfunc/)
-  documentation (http://corrfunc.rtfd.io/)   
-  version control (https://github.com/manodeep/corrfunc)

.. |Release| image:: https://img.shields.io/github/release/manodeep/corrfunc.svg
   :target: https://github.com/manodeep/corrfunc/releases/latest
   :alt: Latest Release

.. |PyPI| image:: https://img.shields.io/pypi/v/corrfunc.svg
   :target: https://pypi.python.org/pypi/corrfunc
   :alt: PyPI Release
.. |MIT licensed| image:: https://img.shields.io/badge/license-MIT-blue.svg
   :target: https://raw.githubusercontent.com/manodeep/corrfunc/master/LICENSE
   :alt: MIT License
.. |ASCL| image:: https://img.shields.io/badge/ascl-1703.003-blue.svg?colorB=262255
   :target: http://ascl.net/1703.003
   :alt: ascl:1703.003
.. |Travis Build| image:: https://travis-ci.org/manodeep/Corrfunc.svg?branch=master
   :target: https://travis-ci.org/manodeep/Corrfunc
   :alt: Build Status
.. |Issues| image:: https://img.shields.io/github/issues/manodeep/corrfunc.svg
   :target: https://github.com/manodeep/corrfunc/issues
   :alt: Open Issues
.. |RTD| image:: https://readthedocs.org/projects/corrfunc/badge/?version=master
   :target: http://corrfunc.readthedocs.io/en/master/?badge=master
   :alt: Documentation Status
.. |Landscape| image:: https://landscape.io/github/manodeep/corrfunc/master/landscape.svg?style=flat
   :target: https://landscape.io/github/manodeep/corrfunc/master
   :alt: Code Health



CHANGELOG.rst
2.0.0 (upcoming)
=================

New features
------------

- Library behaviour can now be controlled at runtime
- Calculates with ``doubles`` and ``floats`` transparently
  (passing arrays of ``doubles`` ensures calculation in double
  precision)
- Both the API and ABI should be future proof
- Extensive docs (first version with docs)
- Arbitrary cosmology can be accounted for in the mocks routines  `#71 <https://github.com/manodeep/corrfunc/issues/71>`_
  
**Breaking Changes**
---------------------

- API has changed from previous version. Two additional inputs are
  now required for every statistic (`#73 <https://github.com/manodeep/corrfunc/issues/73>`_)
  

Enhancements
------------

- Ctrl-C now aborts even within python extensions (cleans up memory too!, `#12 <https://github.com/manodeep/corrfunc/issues/12>`_)
- Significantly improved installation for python
  - compiler can now be specified within ``python setup.py install
    CC=yourcompiler`` `#31<https://github.com/manodeep/corrfunc/issues/31>`_
  - python via an alias is now solved `#52 <https://github.com/manodeep/corrfunc/issues/52>`_


Bug fixes
----------

- Fixed bug in ``DDrppi_mocks`` where the minimum number of grid cells had to
  be 1 `#70 <https://github.com/manodeep/corrfunc/issues/70>`_
  


Outstanding issues
-------------------
- Conda package still is pending (`#49 <https://github.com/manodeep/corrfunc/issues/49>`_)
- Recursive Makefile needs to be replaced with
  a more monolithic Makefile (`#14 <https://github.com/manodeep/corrfunc/issues/14>`_)
- Parameter parsing in python extensions can be flaky (`#79 <https://github.com/manodeep/corrfunc/issues/79>`_)


1.1.0 (June 8, 2016)
=====================

- SSE kernels for all statistics
- Incorrect normalization in ``xi``. **ALL** previous
  ``xi`` calculations were wrong.


1.0.0 (Apr 14, 2016)
====================

- Improved installation process  
- Detecting ``AVX`` capable CPU at compile time
- Double-counting bug fixes in ``wp`` and ``xi``
  

0.2.3 (Mar 30, 2016)
=====================

- Streamlined compilation on MACs
- PyPI version is not verbose by default


0.2.2 (Feb 9, 2016)
====================

- First version on `PyPI <https://pypi.python.org/pypi/corrfunc>`_


0.2.1 (Feb 6, 2016)
====================

- ``AVX`` enabled by default


0.2.0 (Feb 5, 2016)
====================

- Python 2/3 compatible
 


0.0.1 (Nov 11, 2015)
====================

- Initial release



utils/defs.h
/* File: defs.h */
/*
  This file is a part of the corrfunc package
  Copyright (C) 2015-- Manodeep Sinha (manodeep@gmail.com)
  License: MIT LICENSE. See LICENSE file under the top-level
  directory at https://github.com/manodeep/corrfunc/
*/

#pragma once

#include <stdio.h>
#include <math.h>
#include <string.h>
#include <inttypes.h>

#include "macros.h"

#ifdef __cplusplus
extern "C" {
#endif

#define API_VERSION          STR("2.0.1")

typedef enum {
  DEFAULT=-42,/* present simply to make the enum a signed int*/
  FALLBACK=0, /* No special options */
  SSE=1,  /* 64 bit vectors */
  SSE2=2, /* 128 bit vectors */
  SSE3=3, /* 128 bit vectors */
  SSSE3=4, /* 128 bit vectors */
  SSE4=5,/* 128bit vectors */
  SSE42=6, /* 128bit vectors with blend operations */
  AVX=7, /* 256bit vector width */
  AVX2=8,  /* AVX2 (integer operations)*/
  AVX512F=9,/* AVX 512 Foundation */
  NUM_ISA  /*NUM_ISA will be the next integer after 
            the last declared enum. AVX512F:=9 (so, NUM_ISA==10)*/
} isa;//name for instruction sets -> corresponds to the return from instrset_detect in cpu_features.c


/* Macros as mask for the binning_flags */
/* These consititute the 32 bytes for 
the ``uint32_t binning_flags`` */
    
#define BINNING_REF_MASK         0x0000000F //Last 4 bits for how the bin sizes are calculated is done. Also indicates if refines are in place
#define BINNING_ORD_MASK         0x000000F0 //Next 4 bits for how the 3-D-> 1-D index conversion
/* The upper 24 bits are unused currently */

#define BINNING_DFL   0x0
#define BINNING_CUST  0x1
    
struct api_cell_timings
{
    int64_t N1;/* Number of points in the first cell*/
    int64_t N2;/* Number of points in the second cell */
    int64_t time_in_ns;/* Time taken in the compute kernel, measured in nano-seconds*/
    int first_cellindex;
    int second_cellindex;
    int tid;/* Thread-id, 0 for serial case, wastes 4 bytes, since thread id is 4bytes integer and not 8 bytes */
};
    

#define OPTIONS_HEADER_SIZE     (1024)
struct config_options
{
    /* The fields should appear here in decreasing order of 
       alignment requirements. Generally speaking, alignment
       is at least the sizeof the variable type. double has
       8 byte alignment, int has 4 bytes, char has 1 byte etc...
       (size_t could be 4 or 8 bytes depending on compilation 
       mode)
     */

    /* Theory option for periodic boundaries */
    double boxsize;

    /* Options for mocks */
    //cosmology struct. Intentionally left anoynoymous, so I can
    //directly access the fields.
    struct{
        double OMEGA_M;
        double OMEGA_B;
        double OMEGA_L;
        double HUBBLE;
        double LITTLE_H;
        double SIGMA_8;
        double NS;
    };

    /* Measures the time spent in the C API while accessed from python.
       Enabled when the flag c_timer is set
     */
    double c_api_time;

    /* Per cell timers. Keeps track of the number of particles per cell pair
       and time spent to compute the pairs. Might slow down code */
    struct api_cell_timings *cell_timings;
    int64_t totncells_timings;
    
    
    size_t float_type; /* floating point type -> vectorized supports double/float; fallback can support long double*/
    int32_t instruction_set; /* select instruction set to run on */

    char version[32];/* fill in the version number */
    uint8_t verbose; /* Outputs progressbar and times */
    uint8_t c_api_timer; /* Measures time spent in the C function */
    uint8_t c_cell_timer;/* Measures time spent per cell-pair. Might slow down the code */

    /* Options valid for both theory and mocks */
    uint8_t need_avg_sep; /* <rp> or <\theta> is required */
    uint8_t autocorr;/* Only one dataset is required */
    
    /* Options for theory*/
    uint8_t periodic; /* count in periodic mode? flag ignored for wp/xi */
    uint8_t sort_on_z;/* option to sort particles based on their Z co-ordinate in gridlink*/

    /* For DDrppi_mocks and vpf*/
    uint8_t is_comoving_dist;/* flag to indicate cz is already co-moving distance */
    
    /* the link_in_* variables control how the 3-D cell structure is created */
    uint8_t link_in_dec;/* relevant for DDthteta_mocks */
    uint8_t link_in_ra; /* relevant for DDtheta_mocks.*/

    /* Replaces the divide in DDrppi_mocks in AVX mode by a reciprocal and a Newton-Raphson step. */
    uint8_t fast_divide;//Only used in AVX

    /* Fast arccos for wtheta (effective only when OUTPUT_THETAAVG is enabled) */
    uint8_t fast_acos;


    int8_t bin_refine_factors[3];/* Array for the custom bin refine factors in each dim 
                                   xyz for theory routines and ra/dec/cz for mocks
                                   Must be signed integers since some for loops might use -bin_refine_factor
                                   as the starting point */

    uint16_t max_cells_per_dim;/* max number of cells per dimension. same for both theory and mocks */
    union{
        uint32_t binning_flags;/* flag for all linking features, 
                                  Will contain OR'ed flags from enum from `binning_scheme`
                                  Intentionally set as unsigned int, since in the 
                                  future we might want to support some bit-wise OR'ed
                                  functionality */
        uint8_t bin_masks[4];
    };

    /* Reserving to maintain ABI compatibility for the future */
    /* Note that the math here assumes no padding bytes, that's because of the 
       order in which the fields are declared (largest to smallest alignments)  */
    uint8_t reserved[OPTIONS_HEADER_SIZE - 33*sizeof(char) - sizeof(size_t) - 9*sizeof(double) - 3*sizeof(int)
                     - sizeof(uint16_t) - 14*sizeof(uint8_t) - sizeof(struct api_cell_timings *) - sizeof(int64_t) ];
};

static inline void set_bin_refine_scheme(struct config_options *options, const int8_t flag)
{
    //Set the top (nbits-4) to whatever already exists in binning_flag
    //and then set the bottom 4 bits to BIN_DFL
    options->binning_flags = (options->binning_flags & ~BINNING_REF_MASK) | (flag & BINNING_REF_MASK);
}

    
static inline void reset_bin_refine_scheme(struct config_options *options)
{
    set_bin_refine_scheme(options, BINNING_DFL);
}

static inline int8_t get_bin_refine_scheme(struct config_options *options)
{
    //Return the last 4 bits as 8 bits int
    return (int8_t) (options->binning_flags & BINNING_REF_MASK);
}
    
static inline void set_bin_refine_factors(struct config_options *options, const int bin_refine_factors[3])
{
    for(int i=0;i<3;i++) {
        int8_t bin_refine = bin_refine_factors[i];
        if(bin_refine_factors[i] > INT8_MAX) {
            fprintf(stderr,"Warning: bin refine factor[%d] can be at most %d. Found %d instead\n", i,
                    INT8_MAX, bin_refine_factors[i]);
            bin_refine = 1;
        }
        options->bin_refine_factors[i] = bin_refine;
        
    }
    /*
      Note, programmatically setting the refine factors resets the binning flag to "BINNING_DFL"
      BINNING_CUST is only set via function parameters, or explicitly */
    reset_bin_refine_scheme(options);
}

static inline void reset_bin_refine_factors(struct config_options *options)
{
    /* refine factors of 2,2,1 in the xyz dims
       seems to produce the fastest code */
    options->bin_refine_factors[0] = 2;
    options->bin_refine_factors[1] = 2;
    options->bin_refine_factors[2] = 1;
    reset_bin_refine_scheme(options);
}

    

static inline void set_max_cells(struct config_options *options, const int max)
{
    if(max <=0) {
        fprintf(stderr,"Warning: Max. cells per dimension was requested to be set to "
                "a negative number = %d...returning\n", max);
        return;
    }

    if(max > INT16_MAX) {
        fprintf(stderr,"Warning: Max cells per dimension is a 2-byte integer and can not "
                "hold supplied value of %d. Max. allowed value for max_cells_per_dim is %d\n",
                max, INT16_MAX);
    }
    
    options->max_cells_per_dim = max;
}

static inline void reset_max_cells(struct config_options *options)
{
    options->max_cells_per_dim = NLATMAX;
}
    

static inline struct config_options get_config_options(void)
{
    ENSURE_STRUCT_SIZE(struct config_options, OPTIONS_HEADER_SIZE);//compile-time check for making sure struct is correct size

    if(strncmp(API_VERSION, STR(VERSION), 32) != 0) {
        fprintf(stderr,"Error: Version mismatch between header and Makefile. Header claims version = `%s' while Makefile claims version = `%s'\n"
                "Library header probably needs to be updated\n", API_VERSION, STR(VERSION));
        exit(EXIT_FAILURE);
    }
    struct config_options options;
    BUILD_BUG_OR_ZERO(sizeof(options.max_cells_per_dim) == sizeof(int16_t), max_cells_per_dim_must_be_16_bits);
    BUILD_BUG_OR_ZERO(sizeof(options.binning_flags) == sizeof(uint32_t), binning_flags_must_be_32_bits);
    BUILD_BUG_OR_ZERO(sizeof(options.bin_refine_factors[0]) == sizeof(int8_t), bin_refine_factors_must_be_8_bits);
    
    memset(&options, 0, OPTIONS_HEADER_SIZE);
    snprintf(options.version, sizeof(options.version)/sizeof(char)-1, "%s", API_VERSION);
#ifdef DOUBLE_PREC
    options.float_type = sizeof(double);
#else
    options.float_type = sizeof(float);
#endif 
#ifndef SILENT
    options.verbose = 1;
#endif
    
#ifdef OUTPUT_RPAVG    
    options.need_avg_sep = 1;
#endif
#ifdef PERIODIC
    options.periodic = 1;
#endif    

#ifdef __AVX__
    options.instruction_set = AVX;
#elif defined(__SSE4_2__)
    options.instruction_set = SSE42;
#else
    options.instruction_set = FALLBACK;
#endif

    /* Options specific to mocks */

    /* Options for DDrppi_mocks */
#ifdef FAST_DIVIDE
    options.fast_divide=1;
#endif

    
    /* Options for wtheta*/
#ifdef OUTPUT_THETAAVG
    options.need_avg_sep = 1;
#endif

#ifdef LINK_IN_DEC
    options.link_in_dec = 1;
#endif
#ifdef LINK_IN_RA
    options.link_in_ra=1;
    options.link_in_dec=1;
#endif

#ifdef FAST_ACOS
    options.fast_acos=1;
#endif    

#ifdef COMOVING_DIST
    options.is_comoving_dist=1;
#endif

    /* For the thread timings */
    options.totncells_timings = 0;
    /* If the API level timers are requested, then 
       this pointer will have to be allocated */
    options.cell_timings = NULL;
    
    /*Setup the binning options */
    reset_max_cells(&options);
    reset_bin_refine_factors(&options);
    return options;
}


#define EXTRA_OPTIONS_HEADER_SIZE     (1024)

#define MAX_NUM_WEIGHTS 10

typedef struct
{
    void *weights[MAX_NUM_WEIGHTS];  // This will be of shape weights[num_weights][num_particles]
    int64_t num_weights;
} weight_struct;

typedef enum {
  NONE=-42, /* default */
  PAIR_PRODUCT=0,
  NUM_WEIGHT_TYPE 
} weight_method_t; // type of weighting to apply

/* Gives the number of weight arrays required by the given weighting method
 */
static inline int get_num_weights_by_method(const weight_method_t method){
    switch(method){
        case PAIR_PRODUCT:
            return 1;
        default:
        case NONE:
            return 0;
    }
}

/* Maps a name to weighting method
   `method` will be set on return.
 */
static inline int get_weight_method_by_name(const char *name, weight_method_t *method){
    if(name == NULL || strcmp(name, "") == 0){
        *method = NONE;
        return EXIT_SUCCESS;
    }
    // These should not be strncmp because we want the implicit length comparison of strcmp.
    // It is still safe because one of the args is a string literal.
    if(strcmp(name, "pair_product") == 0 || strcmp(name, "p") == 0){
        *method = PAIR_PRODUCT;
        return EXIT_SUCCESS;
    }
        
    return EXIT_FAILURE;
}
    
struct extra_options
{
    // Two possible weight_structs (at most we will have two loaded sets of particles)
    weight_struct weights0;
    weight_struct weights1;
    weight_method_t weight_method; // the function that will get called to give the weight of a particle pair
    uint8_t reserved[EXTRA_OPTIONS_HEADER_SIZE - 2*sizeof(weight_struct) - sizeof(weight_method_t)];
};

// weight_method determines the number of various weighting arrays that we allocate
static inline struct extra_options get_extra_options(const weight_method_t weight_method)
{    
    struct extra_options extra;
    ENSURE_STRUCT_SIZE(struct extra_options, EXTRA_OPTIONS_HEADER_SIZE);//compile-time check for making sure struct is correct size
    memset(&extra, 0, EXTRA_OPTIONS_HEADER_SIZE);
    
    extra.weight_method = weight_method;
    
    weight_struct *w0 = &(extra.weights0);
    weight_struct *w1 = &(extra.weights1);
    w0->num_weights = get_num_weights_by_method(extra.weight_method);
    w1->num_weights = w0->num_weights;

    return extra;
}

static inline void print_cell_timings(struct config_options *options)
{
    fprintf(stderr,"#########################################################################\n");
    fprintf(stderr,"#  Cell_1    Cell_2          N1          N2        Time_ns     ThreadID  \n");
    fprintf(stderr,"#########################################################################\n");
    for(int64_t i=0;i<options->totncells_timings;i++) {
        fprintf(stderr,"%8d %8d %12"PRId64" %12"PRId64" %12"PRId64" %12d\n",
                options->cell_timings[i].first_cellindex,
                options->cell_timings[i].second_cellindex,
                options->cell_timings[i].N1,
                options->cell_timings[i].N2,
                options->cell_timings[i].time_in_ns,
                options->cell_timings[i].tid);
    }

}

static inline void free_cell_timings(struct config_options *options)
{
    if(options->totncells_timings > 0 && options->cell_timings != NULL) {
        free(options->cell_timings);
    }
}    

static inline void assign_cell_timer(struct api_cell_timings *cell_timings, const int64_t totncells, const int max_ngb_cells, struct config_options *options)
{
    int64_t totncells_timings=0;
    for(int64_t index1=0;index1<totncells;index1++) {
        for(int ingb=0;ingb<max_ngb_cells;ingb++) {
            int index = index1 * max_ngb_cells + ingb;
            struct api_cell_timings *t = &(cell_timings[index]);
            if(t->time_in_ns > 0) {
                totncells_timings++;
            }
        }
    }
    /* Does the existing thread timings pointer have enough memory allocated ?*/
    if(options->totncells_timings < totncells_timings) {
        options->totncells_timings = 0;
        
        /* Not enough memory -> need to reallocate*/
        free(options->cell_timings);
        options->cell_timings = calloc(totncells_timings, sizeof(*(options->cell_timings)));
        if(options->cell_timings == NULL) {
            fprintf(stderr,"Warning: In %s> Could not allocate memory to store the API timings per cell. \n",
                    __FUNCTION__);
        } else {
            options->totncells_timings = totncells_timings;
        }
    } 
    
    /* This looks like a repeated "if" condition but it is not. Covers the case for the calloc failure above */
    if(options->totncells_timings >= totncells_timings) {
        /* Okay, enough memory to assign the thread timings */
        struct api_cell_timings *t = options->cell_timings;
        for(int64_t index1=0;index1<totncells;index1++) {
            for(int ingb=0;ingb<max_ngb_cells;ingb++) {
                int index = index1 * max_ngb_cells + ingb;
                if(cell_timings[index].time_in_ns > 0) {
                    *t = cell_timings[index];
                    t++;
                }
            }
        }
    }
}
    
    
#include "macros.h"

    
#ifdef __cplusplus
}
#endif
